lib/vdsm/virt/vm.py

```diff
@@ -1782,6 +1782,11 @@ def setDownStatus(self, code, exitReasonCode, exitMessage=''):
                 self.guestAgent.stop()
             except Exception:
                 pass
+            finally:
+                if self._dom.connected:
+                    self.log.info("Switching domain %r to Defined", self.id)
+                    dom = self._connection.lookupByUUIDString(self.id)
```

Comment:
I didn't quite get what is the risk of using `self._dom`. If this is just about possibly wrapping existing virdomain object then maybe having something like this here or in `virdomain.*.__init__` would help:

```
if hasattr(dom, "dom"):
    self._dom = dom.dom
else: 
    self._dom = dom
````

If there is also a problem that the domain could get undefined (by libvirt?) in the mean time then the current code is incomplete and and we need to handle error from `lookupByUUIDString()`. Otherwise I don't see a benefit in using it.

Reply:
I'm not sure we understand each other; my point was possibly avoiding the libvirt call here and use something like `dom = self._dom.dom` here (we check for `connected` above so `self._dom.dom` should be available). If we liked to be super-robust we could do something like

``` python
try:
    dom = self._dom.dom
except AttributeError:
    try:
        dom = self._connection.lookupByUUIDString(self.id)
    except:
        dom = None  # ?
self._dom = virdomain.Defined(self.id, dom)
```

Reply:
> I'm not sure we understand each other;

I think we do. :)

> we check for `connected`

You are right. This seems good enough to me.